File: donghwa/arxiv/MST/1584_kruskal.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.parent[root_x] = root_y

def minCostConnectPoints(points):
    n = len(points)
    edges = []  # Store edges in the format (cost, point1, point2)
    
    for i in range(n):
        for j in range(i + 1, n):
            manhattan_dist = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1])
            edges.append((manhattan_dist, i, j))
    
    # Sort edges by their cost
    edges.sort()
    
    uf = UnionFind(n)
    min_cost = 0
    edges_added = 0
    
    for edge in edges:
        cost, u, v = edge
        if uf.find(u) != uf.find(v):
            uf.union(u, v)
            min_cost += cost
            edges_added += 1
            if edges_added == n - 1:
                break  # Minimum spanning tree is complete
    
    return min_cost

# Test the function with the given example
points = [[0,0],[2,2],[3,10],[5,2],[7,0]]
print(minCostConnectPoints(points))  # Output should be 20

File: donghwa/arxiv/MST/1584_kruskal.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Define a structure to represent edges
struct Edge {
    int src, dest, weight;
};

// Define a structure to represent a disjoint-set (Union-Find)
class UnionFind {
public:
    vector<int> parent;
    
    UnionFind(int n) {
        parent.resize(n);
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }
    
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    void unionSets(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            parent[rootX] = rootY;
        }
    }
};

// Compare function for sorting edges by weight
bool compareEdges(const Edge& a, const Edge& b) {
    return a.weight < b.weight;
}

int minCostConnectPoints(vector<vector<int>>& points) {
    int n = points.size();
    
    // Create an array of edges
    int numEdges = n * (n - 1) / 2;
    vector<Edge> edges(numEdges);
    int edgeIndex = 0;
    
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            int manhattanDist = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]);
            edges[edgeIndex].src = i;
            edges[edgeIndex].dest = j;
            edges[edgeIndex].weight = manhattanDist;
            edgeIndex++;
        }
    }
    
    // Sort edges by weight
    sort(edges.begin(), edges.end(), compareEdges);
    
    // Initialize Union-Find data structure
    UnionFind uf(n);
    
    int minCost = 0;
    int edgesAdded = 0;
    
    for (int i = 0; i < numEdges; i++) {
        int src = edges[i].src;
        int dest = edges[i].dest;
        int weight = edges[i].weight;
        
        if (uf.find(src) != uf.find(dest)) {
            uf.unionSets(src, dest);
            minCost += weight;
            edgesAdded++;
            if (edgesAdded == n - 1) {
                break;  // Minimum spanning tree is complete
            }
        }
    }
    
    return minCost;
}

int main() {
    vector<vector<int>> points = {{0, 0}, {2, 2}, {3, 10}, {5, 2}, {7, 0}};
    
    int result = minCostConnectPoints(points);
    cout << "Minimum cost: " << result << endl;  // Output should be 20
    
    return 0;
}
